lib/mongo/cluster.rb

```diff
@@ -153,9 +151,12 @@ def elect_primary!(description)
     def remove(host)
       log_debug([ "#{host} being removed from the cluster." ])
       address = Address.new(host)
-      removed_servers = @servers.reject!{ |server| server.address == address }
+      removed_servers = @servers.select { |s| s.address == address }
+      @servers_update.synchronize { @servers = @servers - removed_servers }
       removed_servers.each{ |server| server.disconnect! } if removed_servers
-      addresses.reject!{ |addr| addr == address }
+      @addresses_update.synchronize do
+        @addresses.reject!{ |addr| addr == address }
+      end
```

Comment:
Is it ok that another thread can see addresses and servers go out of sync between these statements? If you just use one lock you can avoid answering that question. =)